pkg/visitors/table.go

```diff
@@ -26,57 +31,105 @@ func (v *Table) Run(f uefi.Firmware) error {
 
 // Visit applies the Table visitor to any Firmware type.
 func (v *Table) Visit(f uefi.Firmware) error {
+	var offset uint64
 	switch f := f.(type) {
 	case *uefi.FlashImage:
-		return v.printRow(f, "Image", "", "")
+		v.depth = v.indent + 1
+		return v.printRow(f, "Image", "", "", 0, 0)
 	case *uefi.FirmwareVolume:
-		return v.printRow(f, "FV", f.FileSystemGUID.String(), "")
+		return v.printRow(f, "FV", f.FileSystemGUID.String(), "", v.offset+f.FVOffset, v.offset+f.FVOffset+f.DataOffset)
 	case *uefi.File:
 		// TODO: make name part of the file node
-		return v.printRow(f, "File", f.Header.GUID.String(), f.Header.Type)
+		return v.printRow(f, "File", f.Header.GUID.String(), f.Header.Type, v.curOffset, v.curOffset+f.DataOffset)
 	case *uefi.Section:
-		return v.printRow(f, "Sec", f.String(), f.Type)
+		// Reset offset to O for (compressed) section content
+		return v.printRow(f, "Sec", f.String(), f.Type, v.curOffset, 0)
 	case *uefi.FlashDescriptor:
-		return v.printRow(f, "IFD", "", "")
+		v.depth = v.indent + 1
+		return v.printRow(f, "IFD", "", "", 0, 0)
 	case *uefi.BIOSRegion:
-		return v.printRow(f, "BIOS", "", "")
+		v.depth = v.indent + 1
+		if f.FRegion != nil {
+			offset = uint64(f.FRegion.BaseOffset())
+		}
+		return v.printRow(f, "BIOS", "", "", offset, offset)
 	case *uefi.BIOSPadding:
-		return v.printRow(f, "BIOS Pad", "", "")
+		return v.printRow(f, "BIOS Pad", "", "", v.offset+f.Offset, 0)
 	case *uefi.NVarStore:
-		return v.printRow(f, "NVAR Store", "", "")
+		return v.printRow(f, "NVAR Store", "", "", v.curOffset, v.curOffset)
 	case *uefi.NVar:
-		return v.printRow(f, "NVAR", f.GUID.String(), f)
+		return v.printRow(f, "NVAR", f.GUID.String(), f, v.curOffset, v.curOffset+uint64(f.DataOffset))
 	case *uefi.RawRegion:
-		return v.printRow(f, f.Type().String(), "", "")
+		v.depth = v.indent + 1
+		if f.FRegion != nil {
+			offset = uint64(f.FRegion.BaseOffset())
+		}
+		return v.printRow(f, f.Type().String(), "", "", offset, offset)
 	default:
-		return v.printRow(f, fmt.Sprintf("%T", f), "", "")
+		return v.printRow(f, fmt.Sprintf("%T", f), "", "", 0, 0)
 	}
 }
 
 func indent(n int) string {
 	return strings.Repeat(" ", n)
 }
 
-func (v *Table) printRow(f uefi.Firmware, node, name, typez interface{}) error {
+func (v *Table) printRow(f uefi.Firmware, node, name, typez interface{}, offset, dataOffset uint64) error {
```

Comment:
I highly recommend a refactor here. A large part of this code is dependent on if v.Layout vs otherwise.

We can adapt this for different table styles by making two printRow functions, and keeping a variable inside the table visitor struct 

type Table struct {
	W      *tabwriter.Writer
	indent int
	Layout    bool
	TopLevel  bool
	indent    int
	offset    uint64
	curOffset uint64
	depth     int
	print       func(f uefi.Firmware, node, name, typez interface{}, offset, dataOffset uint64) error 
}

and assigning the print based on whether Layout is set.

For even more code reuse, we can split the print function into two segments instead, one which does all the printing prior to applychildren called printData, and one which does all the printing after applychildren called printFooter, and this printRow would call v.printData(...) then f.ApplyChildren then v.printFooter(...)

Reply:
I think this would make it much clearer to the reader.

Reply:
I did a first refactoring.
I'm bothered with the function in the struct... It can't use the receiver anymore, right ? So I'd have to give `v *Table` as an argument

Reply:
done